core/iwasm/libraries/libc-wasi/libc_wasi_wrapper.c

```diff
@@ -969,9 +988,36 @@ wasi_poll_oneoff(wasm_exec_env_t exec_env, const wasi_subscription_t *in,
         || !validate_native_addr(nevents_app, sizeof(uint32)))
         return (wasi_errno_t)-1;
 
-    err = wasmtime_ssp_poll_oneoff(curfds, in, out, nsubscriptions, &nevents);
-    if (err)
-        return err;
+    __wasi_timestamp_t elapsed = 0;
+    /* in seconds */
+    const __wasi_timestamp_t timeout = get_timeout_for_poll_oneoff(in),
+                             time_quant = 1;
+    const uint32 size_to_copy = nsubscriptions * sizeof(wasi_subscription_t);
+    __wasi_subscription_t *in_copy;
+    if (size_to_copy
+        && !(in_copy =
+                 (__wasi_subscription_t *)wasm_runtime_malloc(size_to_copy))) {
+        return (wasi_errno_t)-1;
+    }
+    bh_memcpy_s(in_copy, size_to_copy, in, size_to_copy);
+
+    while (1) {
+        elapsed += time_quant;
+        if ((int)timeout != -1 && elapsed > timeout)
+            break;
+
+        in_copy[0].u.u.clock.timeout = time_quant * 1e9;
```

Comment:
What if zero-th subscription is not a clock subscription (but it's Read or Write)?

Reply:
iterating on subscriptions and updated timeout of all clock subscriptions only instead